File: android/app/src/main/kotlin/in/nann/smashhockey/screens/TrainingScreen.kt
package `in`.nann.smashhockey.screens

import `in`.nann.smashhockey.core.generated.CopyKey
import `in`.nann.smashhockey.core.generated.Drill
import `in`.nann.smashhockey.core.generated.Role
import `in`.nann.smashhockey.core.generated.SaveRecord
import `in`.nann.smashhockey.core.season.isOpen
import `in`.nann.smashhockey.core.season.isWon
import `in`.nann.smashhockey.game.Game
import `in`.nann.smashhockey.game.L
import `in`.nann.smashhockey.game.MatchPlan
import `in`.nann.smashhockey.game.Names
import `in`.nann.smashhockey.generated.Presentation
import `in`.nann.smashhockey.ui.BlockButton
import `in`.nann.smashhockey.ui.Entrance
import `in`.nann.smashhockey.ui.KitSound
import `in`.nann.smashhockey.ui.Label3D
import `in`.nann.smashhockey.ui.Panel
import `in`.nann.smashhockey.ui.Paragraph
import `in`.nann.smashhockey.ui.Presentable
import `in`.nann.smashhockey.ui.Quat
import `in`.nann.smashhockey.ui.Tile
import `in`.nann.smashhockey.ui.WaveText

/**
 * Training (spec §16.6, §10) — the twin of iOS's TrainingScreen.swift: the eight drills as cards
 * in order — name, world, goals and time, what opposes, won or locked. A locked card shakes its
 * head. Choosing an open one tips its intro card up — the hint — with Start, which goes straight to
 * the drill's "get ready".
 */
class TrainingScreen(game: Game, intro: Drill?) :
    Screen(Game.pose(Presentation.Screens.Training.eye, Presentation.Screens.Training.target), game) {
    private var introParts: List<Presentable> = emptyList()
    private val cards = ArrayList<Tile>()

    init {
        val save = game.save
        part(WaveText(kit, L(CopyKey.TRAINING_TITLE), 0.17f, C.PAPER, bob = 0.6f, id = "training_title_header"), at(0f, top - 0.2f))
        val done = Drill.entries.count { save.isWon(it) }
        part(Label3D(kit, L(CopyKey.TRAINING_PROGRESS, done, Drill.entries.size), S.TEXT_SMALL, C.SUN, maxWidth = 1.6f,
            entrance = Entrance.Drop), at(0f, top - 0.43f))
        val contentTop = top - 0.58f
        val contentBottom = bottom + 0.62f
        val pitch = minOf(0.5f, (contentTop - contentBottom) / 4)
        for ((i, drill) in Drill.entries.withIndex()) {
            val x = if (i % 2 == 0) -0.45f else 0.45f
            val y = contentTop - pitch * (i / 2 + 0.5f)
            cards += part(card(drill, pitch - 0.05f, save), at(x, y))
        }
        part(BlockButton(kit, L(CopyKey.COMMON_BACK), "training_back_button", BlockButton.Style.QUIET, 0.8f, 0.3f, 0.09f) {
            game.go(game.home)
        }, at(0f, bottom + 0.3f))
        if (intro != null && save.isOpen(intro)) stage.after(0.9) { open(intro) }
    }

    private fun card(d: Drill, h: Float, save: SaveRecord): Tile {
        val open = save.isOpen(d)
        val won = save.isWon(d)
        val state = if (won) L(CopyKey.TRAINING_WON) else if (open) "" else L(CopyKey.TRAINING_LOCKED)
        val label = listOf(L(CopyKey.TRAINING_DRILL, d.number), L(d.nameKey), L(d.world.nameKey), state).filter { it.isNotEmpty() }.joinToString(", ")
        val t = Tile(kit, 0.86f, h, if (won) C.ROW_DARK else C.PAPER, id = "training_drill_${d.number}_button", label = label) { open(d) }
        t.isEnabled = open
        val ink = if (open) C.INK else C.DISABLED_INK
        letters(L(CopyKey.TRAINING_DRILL, d.number), 0.032f, if (open) C.GREEN_INK else C.DISABLED_INK, align = Label3D.Align.LEADING,
            x = -0.39f, y = h * 0.3f, parent = t.content)
        letters(L(d.nameKey).uppercase(), 0.05f, ink, maxWidth = 0.76f, align = Label3D.Align.LEADING, x = -0.39f, y = h * 0.08f, parent = t.content)
        val facts = "${Names.world(d.world)} · ${L(CopyKey.TRAINING_GOALS_IN, d.goals, d.seconds.toInt())}"
        letters(facts, 0.03f, ink, maxWidth = 0.76f, align = Label3D.Align.LEADING, x = -0.39f, y = -h * 0.13f, parent = t.content)
        letters(L(opposition(d)), 0.03f, if (open) C.PINK_INK else C.DISABLED_INK, maxWidth = 0.5f, align = Label3D.Align.LEADING,
            x = -0.39f, y = -h * 0.3f, parent = t.content)
        if (won || !open) {
            val badge = kit.slab(0.26f, 0.08f, 0.03f, if (won) C.GREEN else C.DISABLED_SHADE, t.content, corner = 0.03f)
            badge.setPosition(0.27f, -h * 0.3f, 0.02f)
            badge.setRotation(Quat().axisAngle(-0.12f, 0f, 0f, 1f))
            letters(state, 0.035f, C.INK, maxWidth = 0.22f, x = 0.27f, y = -h * 0.3f, z = 0.04f, parent = t.content)
        }
        return t
    }

    /** The intro card (§16.4, §16.6): the drill's name, its hint, and Start. */
    private fun open(d: Drill) {
        if (!game.save.isOpen(d)) return
        closeIntro()
        for ((i, c) in cards.withIndex()) c.isSelected = Drill.entries[i] == d
        cards.forEach { it.isEnabled = false }      // the cards behind the intro take no taps (§16.6)
        val panel = child(Panel(kit, 1.62f, 1.5f, 0.14f, C.PAPER, Entrance.Tumble), at(0f, 0.05f, z = 0.5f), layer)
        child(Label3D(kit, L(CopyKey.TRAINING_DRILL, d.number), 0.05f, C.GREEN_INK), at(0f, 0.6f), panel.content).show(0.0)
        child(Label3D(kit, L(d.nameKey).uppercase(), 0.11f, C.INK, maxWidth = 1.45f), at(0f, 0.45f), panel.content).show(0.0)
        child(Label3D(kit, L(CopyKey.TRAINING_GOALS_IN, d.goals, d.seconds.toInt()), 0.05f, C.PINK_INK, maxWidth = 1.4f),
            at(0f, 0.31f), panel.content).show(0.0)
        child(Paragraph(kit, L(d.hintKey), 0.058f, C.INK, 1.4f, id = "training_hint"), at(0f, -0.02f), panel.content).show(0.0)
        val start = child(BlockButton(kit, L(CopyKey.TRAINING_START), "training_start_button", BlockButton.Style.PRIMARY, 0.8f, 0.32f, 0.13f) {
            game.play(MatchPlan.Practice(d))
        }, at(0.33f, -0.52f), panel.content)
        start.bobs = true
        val close = child(BlockButton(kit, L(CopyKey.COMMON_BACK), "training_close_button", BlockButton.Style.QUIET, 0.56f, 0.28f, 0.08f) {
            closeIntro()
        }, at(-0.46f, -0.52f), panel.content)
        val parts = listOf(panel, start, close)
        KitSound.sweep()
        for ((i, p) in parts.withIndex()) p.show(i * motion.staggerSeconds * 3)
        introParts = parts
    }

    private fun closeIntro() {
        val panel = introParts.firstOrNull() as? Panel ?: return
        for (p in introParts.asReversed()) p.hide(0.0)
        introParts = emptyList()
        for (c in cards) c.isSelected = false
        for ((i, c) in cards.withIndex()) c.isEnabled = game.save.isOpen(Drill.entries[i])
        stage.after(0.9) { stage.remove(panel.node) }
    }

    override fun leave() {
        for (p in introParts) p.hide(0.0)
        super.leave()
    }

    companion object {
        /** What a drill's opponents are (§10): none, a goalie, dummies, or defenders. */
        fun opposition(d: Drill): CopyKey {
            val roles = d.away.map { it.role }
            return when {
                roles.isEmpty() -> CopyKey.TRAINING_NONE
                roles.any { it == Role.DEFENDER || it == Role.FORWARD } -> CopyKey.TRAINING_DEFENDERS
                Role.DUMMY in roles -> CopyKey.TRAINING_DUMMIES
                else -> CopyKey.TRAINING_GOALIE
            }
        }
    }
}

File: ios/Sources/Screens/TrainingScreen.swift
import RealityKit
import SmashCore

/// Training (spec §16.6, §10): the eight drills as cards in order — name, world, goals and time,
/// what opposes, won or locked. A locked card shakes its head. Choosing an open one tips its intro
/// card up — the hint — with Start, which goes straight to the drill's "get ready". The twin of
/// Android's TrainingScreen.kt.
@MainActor
final class TrainingScreen: Screen {
    private var introParts: [Presentable] = []
    private var cards: [Tile] = []
    private var intro: Drill?

    init(game: Game, intro: Drill?) {
        super.init(pose: CameraPose(Presentation.Screens.Training.eye, Presentation.Screens.Training.target), game: game)
        let m = motion
        let save = game.save
        part(WaveText(L(.trainingTitle), height: 0.17, colour: C.cream, bob: 0.6, id: "training_title_header", motion: m),
             at: at(0, top - 0.2))
        let done = Drill.allCases.filter(save.isWon).count
        part(Label3D(L(.trainingProgress, done, Drill.allCases.count), height: S.textSmall, colour: C.sun, maxWidth: 1.6,
                     entrance: .drop, motion: m), at: at(0, top - 0.43))
        let contentTop = top - 0.58, contentBottom = bottom + 0.62
        let pitch = min(0.5, (contentTop - contentBottom) / 4)
        for (i, drill) in Drill.allCases.enumerated() {
            let x: Float = i % 2 == 0 ? -0.45 : 0.45
            let y = contentTop - pitch * (Float(i / 2) + 0.5)
            cards.append(part(card(drill, height: pitch - 0.05, save: save), at: at(x, y)))
        }
        part(BlockButton(L(.commonBack), id: "training_back_button", style: .quiet, size: [0.8, 0.3], textHeight: 0.09,
                         motion: m) { [weak game] in game.map { $0.go($0.home) } }, at: at(0, bottom + 0.3))
        if let intro, save.isOpen(intro) {
            stage.after(0.9) { [weak self] in self?.open(intro) }
        }
    }

    /// What a drill's opponents are (§10): none, a goalie, dummies, or defenders.
    static func opposition(_ d: Drill) -> CopyKey {
        let roles = d.away.map(\.role)
        if roles.isEmpty { return .trainingNone }
        if roles.contains(where: { $0 == .defender || $0 == .forward }) { return .trainingDefenders }
        if roles.contains(.dummy) { return .trainingDummies }
        return .trainingGoalie
    }

    private func card(_ d: Drill, height h: Float, save: SaveRecord) -> Tile {
        let open = save.isOpen(d), won = save.isWon(d)
        let state = won ? L(.trainingWon) : open ? "" : L(.trainingLocked)
        let label = [L(.trainingDrill, d.number), L(d.nameKey), L(d.world.nameKey), state].filter { !$0.isEmpty }.joined(separator: ", ")
        let t = Tile(size: [0.86, h], colour: won ? C.rowDark : C.cream, id: "training_drill_\(d.number)_button", label: label,
                     motion: motion) { [weak self] in self?.open(d) }
        t.isEnabled = open
        let ink = open ? C.ink : C.disabledInk
        letters(L(.trainingDrill, d.number), height: 0.032, colour: open ? C.tealShade : C.disabledInk, align: .leading,
                at: [-0.39, h * 0.3, 0], on: t.content)
        letters(L(d.nameKey).uppercased(), height: 0.05, colour: ink, maxWidth: 0.76, align: .leading, at: [-0.39, h * 0.08, 0],
                on: t.content)
        let facts = "\(Names.world(d.world)) · \(L(.trainingGoalsIn, d.goals, Int(d.seconds)))"
        letters(facts, height: 0.03, colour: ink, maxWidth: 0.76, align: .leading, at: [-0.39, -h * 0.13, 0], on: t.content)
        letters(L(Self.opposition(d)), height: 0.03, colour: open ? C.coralShade : C.disabledInk, maxWidth: 0.5, align: .leading,
                at: [-0.39, -h * 0.3, 0], on: t.content)
        if won || !open {
            let badge = Blocks.slab([0.26, 0.08, 0.03], won ? C.teal : C.disabledShade, corner: 0.03)
            badge.position = [0.27, -h * 0.3, 0.02]
            badge.orientation = simd_quatf(angle: -0.12, axis: [0, 0, 1])
            t.content.addChild(badge)
            letters(state, height: 0.035, colour: C.cream, maxWidth: 0.22, at: [0.27, -h * 0.3, 0.04], on: t.content)
        }
        return t
    }

    /// The intro card (§16.4, §16.6): the drill's name, its hint, and Start.
    private func open(_ d: Drill) {
        guard game.save.isOpen(d) else { return }
        closeIntro()
        intro = d
        for (i, c) in cards.enumerated() {
            c.isSelected = Drill.allCases[i] == d
            c.isEnabled = false                     // the intro stands in front: the cards behind take no taps
        }
        KitSound.sweep()
        let m = motion
        let panel = child(Panel(size: [1.62, 1.5, 0.14], colour: C.cream, entrance: .tumble, motion: m), at: at(0, 0.05, z: 0.5),
                          on: layer)
        var parts: [Presentable] = [panel]
        child(Label3D(L(.trainingDrill, d.number), height: 0.05, colour: C.tealShade, motion: m), at: at(0, 0.6), on: panel.content)
            .show(after: 0)
        child(Label3D(L(d.nameKey).uppercased(), height: 0.11, colour: C.ink, maxWidth: 1.45, motion: m), at: at(0, 0.45),
              on: panel.content).show(after: 0)
        child(Label3D(L(.trainingGoalsIn, d.goals, Int(d.seconds)), height: 0.05, colour: C.coralShade, maxWidth: 1.4, motion: m),
              at: at(0, 0.31), on: panel.content).show(after: 0)
        let hint = child(Paragraph(L(d.hintKey), height: 0.058, colour: C.ink, width: 1.4, id: "training_hint", motion: m),
                         at: at(0, -0.02), on: panel.content)
        hint.show(after: 0)
        let start = child(BlockButton(L(.trainingStart), id: "training_start_button", style: .primary, size: [0.8, 0.32],
                                      textHeight: 0.13, motion: m) { [weak game] in game?.play(.drill(d)) },
                          at: at(0.33, -0.52), on: panel.content)
        start.bobs = true
        let close = child(BlockButton(L(.commonBack), id: "training_close_button", style: .quiet, size: [0.56, 0.28],
                                      textHeight: 0.08, motion: m) { [weak self] in self?.closeIntro() },
                          at: at(-0.46, -0.52), on: panel.content)
        parts += [start, close]
        for (i, p) in parts.enumerated() { p.show(after: Double(i) * motion.staggerSeconds * 3) }
        introParts = parts
    }

    private func closeIntro() {
        guard !introParts.isEmpty, let panel = introParts.first as? Panel else { return }
        for p in introParts.reversed() { p.hide(after: 0) }
        introParts = []
        intro = nil
        for (i, c) in cards.enumerated() {
            c.isSelected = false
            c.isEnabled = game.save.isOpen(Drill.allCases[i])
        }
        stage.after(0.9) { [weak stage] in stage?.remove(under: panel.entity) }
    }

    override func leave() {
        for p in introParts { p.hide(after: 0) }
        super.leave()
    }
}
